plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/TypingDomain.kt

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.domains
+
+import org.jetbrains.kotlin.formver.embeddings.ClassEmbedding
+import org.jetbrains.kotlin.formver.viper.ast.*
+import org.jetbrains.kotlin.formver.viper.ast.Type.Bool
+
+
+const val TYPE_DOMAIN_NAME = "Type"
+
+/**
+ * Viper Code:
+ * ```
+ * domain Type  {
+ *
+ *   unique function IntType(): Type
+ *   unique function BooleanType(): Type
+ *   unique function UnitType(): Type
+ *   unique function NothingType(): Type
+ *   unique function AnyType(): Type
+ *   unique function FunctionType(): Type
+ *
+ *   function NullableType(t: Type): Type
+ *   function isSubtype(a: Type, b: Type): Bool
+ *   function is_nullable_type(t: Type): Bool
+ *
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != IntType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != BooleanType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != UnitType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != NothingType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != AnyType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != FunctionType())
+ *   }
+ *   // Axioms of the form NullableType(T) != ClassType()
```

Comment:
This comment seems like it should be higher.

Reply:
Here `ClassType()` doesn't refer to the pre-defined types above but instead the user defined classes. So say the user defines a class `Foo` then this is where an axiom get generator with `NullableType(T) != FooType()`

Reply:
Ohh, so this is a TODO?  I think it would be good to mark it as such.

Reply:
No no it's not a TODO, they do get generated but the specific axioms depend on the Kotlin program being verified. So for every class that the user defines in their code there is one type being generated in this domain together with a bunch of axioms.

Reply:
Ahh, right, I see.  Sorry, makes sense.